### PR TITLE
Price handling changes

### PR DESCRIPTION
**Pull request by tjmw**
## What?

Changes to the way ticket pricing information is handled. We can no longer rely on the pricing information returned from the `bookingoptions` response, but instead need to use the new `getPriceMatch` API endpoint (this accepts a list of tickets, and a list of quantities and returns pricing information).

## How?

In the model we now store `Prices` which maps age band IDs to `Price`. When a booking option is selected we make an initial fetch to get the default quantities and the prices, and anytime a quantity is changed we fetch again. We could probably be smart and maybe cache this data locally so we don't have to re-fetch the same data multiple times but time constraints have prevented this.

## Shortcomings

The UI no longer prevents you from setting invalid quantities for different ticket types, since now we have to update the quantity and then check the response to `getPriceMatch` to see if we've got a valid combination (signified by the presence of a non-null total in the response). If we haven't got a valid combination we don't display the total (since we haven't got one) and the button is disabled. In some cases the API doesn't give us the price of a given ticket type (see the example GIF) and we display "..." instead. It sounds like this may be an API issue. There's definitely an API issue where passing certain combinations of quantities to `getPriceMatch` returns an empty response.

We don't do anything in the UI to signal that we're fetching price information, so you may see stale pricing information for a period on a slow connection. In the code we do distinguish between fresh price data and non-fresh, so it would be possible to render things differently.

## Notes

There are several things about the code I'm not super happy about, I think the types I've used could be richer in some cases. I also think German's efforts to make impossible states impossible were worthwhile, and that isn't really the case here. I.e. what does it mean to have ticket pricing information, but no booking option selected. It's possible to represent that, but you can't really get to that state. It also feels like we're leaking more out to the consuming app than we should.

### CODE REVIEW

File: src/BookingWidget/Ticket.elm [hunk 1]
```diff
@@ -1,67 +1,167 @@
...
-type alias Ticket =
-    { price : Price
-    , label : String
+type alias TicketBand =
+    { label : String
     , type_ : String
     , purchaseableQuantities : PurchaseableQuantities
     , ageFrom : Maybe Int
     , ageTo : Maybe Int
+    , priceGroupSortOrder : Int
```
Comment (joshuaclayton) [inline src/BookingWidget/Ticket.elm]: Can this field be removed?

File: src/BookingWidget/Ticket.elm [hunk 2]
```diff
@@ -1,67 +1,167 @@
...
+
+type AgeBandId
+    = AgeBandId Int
+
+
+markPricesAsStale : Prices -> Prices
+markPricesAsStale prices =
+    case prices of
+        FreshPrices total p ->
+            StalePrices total p
```
Comment (joshuaclayton) [inline src/BookingWidget/Ticket.elm]: This is pretty neat!

File: src/BookingWidget/TicketPricingEncoder.elm [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+encodeQuantity ( ageBandId, quantity ) =
+    object
+        [ ( "ageBandId", int <| ageBandIdToInt ageBandId )
+        , ( "quantity", int quantity )
+        ]
+
+
+optionalMax : Maybe Int -> Json.Encode.Value
+optionalMax maybeInt =
+    case maybeInt of
```
Comment (joshuaclayton) [inline src/BookingWidget/TicketPricingEncoder.elm]: `int <| Maybe.withDefault -1 maybeInt`?

File: src/BookingWidget/Update.elm [hunk 4]
```diff
@@ -124,46 +136,92 @@ update msg model =
...
+                            { booking | tickets = Just <| buildDefaultQuantities ticketPricesDocument }
+            in
+                ( { model | ticketPrices = prices, booking = newBooking }, Cmd.none )
+
+        QueryTicketPricingFailure _ ->
+            ( model, Cmd.none )
+
         SelectBookingOption selectedOptionId ->
-            ( selectBookingOption model selectedOptionId, Cmd.none )
+            let
```
Comment (joshuaclayton) [inline src/BookingWidget/Update.elm]: can we remove the `let...in` here and have this call `selectBookingOption model selectedOptionId` only?

File: src/BookingWidget/Update.elm [hunk 5]
```diff
@@ -206,80 +267,65 @@ type TicketOperation
...
+                currentQuantity =
+                    ticketQuantity ageBandId q
 
-        MinimumAndMaximumPurchaseQuantities min max ->
-            n >= min && n <= max
+                updatedQuantity =
+                    case ticketOperation of
+                        IncrementQuantity ->
+                            currentQuantity + 1
```
Comment (joshuaclayton) [inline src/BookingWidget/Update.elm]: This partially happens above when we check if the operation is valid - specifically, taking the operation and current quantity, modifying it (either adding or removing one), and in the case of validation, ensuring the new value is `>= 0`. I wonder if there's something we could extract so these paths don't get traveled multiple times?

File: src/BookingWidget/Ticket.elm (hunk 2 above)
Comment (tjmw) [inline src/BookingWidget/Ticket.elm]: Thanks! We're not _actually_ making use of the difference between fresh and stale prices right now, but this opens up the possibility for doing so in the UI in future.

File: src/BookingWidget/Ticket.elm (hunk 1 above)
Comment (tjmw) [inline src/BookingWidget/Ticket.elm]: Unfortunately, one of the kind of unexpected aspects of the new `getPriceMatch` API endpoint is that you have to pass back the ticket bands you get from the `bookingoptions` endpoint exactly as you receive them. That's the sole reason we have to keep this field around (the same is true of a few others too).

File: src/BookingWidget/TicketPricingEncoder.elm (hunk 3 above)
Comment (tjmw) [inline src/BookingWidget/TicketPricingEncoder.elm]: :+1: Fixed in 0c9ff69e2ea010817c342d5602b32410b78e703a (this also allows us to convert it to point free style)

File: src/BookingWidget/Update.elm (hunk 4 above)
Comment (tjmw) [inline src/BookingWidget/Update.elm]: Good spot, I totally missed that. Fixed in cac6245825ed13ec7ef42a1390ca809b9f1ddb5d

File: src/BookingWidget/Update.elm (hunk 5 above)
Comment (tjmw) [inline src/BookingWidget/Update.elm]: Fixed in `c7f0b5ca07cd6e3bff3cb947ef1d39a3e38d5590`. I've DRYed up the logic of taking a `TicketOperation` and applying it to a quantity. We still call that code twice, but that seems okay to me (command/query separation). What do you think?

**Pull request merged**